### PR TITLE
SV-COMP'24 tracker

### PR DESCRIPTION
**Pull request by fbrausse**
This PR is tracking the svcomp24 branch.

TODO
- [x] merge #1418
- [x] merge #1457
- [x] merge #1464 and set `-C` as default for build.sh
- [x] merge #1463
- [x] merge #1468
- [x] merge #1467 
- [x] merge #1465
- [x] merge #1460
- [x] merge #1450
- [x] merge #1456
- [x] merge #1469
- [x] merge #1428
- [x] merge #1472 and investigate enabling `--compact-trace`; I've started an svcomp24 run with this option to see how much performance we lose: https://github.com/esbmc/esbmc/actions/runs/6853986122
- [x] merge #1466
- [x] merge #1478
- [x] merge #1477
- [x] merge #1482
- [x] merge #1483
- [x] merge #1488
- [x] merge #1489
- [x] merge #1491
- [x] merge #1493
- [x] merge #1496
- [x] merge #1497
- [x] merge #1499 (old version)
- [x] merge #1502
- [x] merge #1500
- [x] merge #1503
- [x] merge #1506 
- [x] merge #1509
- [x] merge #1512
- [x] merge fb/libm-musl-float? [Running](https://github.com/esbmc/esbmc/actions/runs/6919694697), compare to [current](https://github.com/esbmc/esbmc/actions/runs/6919350402)
- [x] decide on:
      - do nothing
      - use [keras2c patch](https://github.com/esbmc/esbmc/commit/17d9fbb3d9e3089e7c933d06f8ea87ac5ea02401); this option was chosen
      - use [Mikhail's patch](https://github.com/esbmc/esbmc/commit/b4ee222aa1d2336f107eddd9740530b609785d2a)
 - [x] merge #1516
 - [x] merge #1519

Open issues:
- [x] #1470
- [x] #1471
- [ ] #1473
- [x] #1484
- [x] #1490
- [x] #1494 part 2
- [x] #1495

Decided against:
- merge #1505 ? [Running](https://github.com/esbmc/esbmc/actions/runs/6919691385), compare to [current](https://github.com/esbmc/esbmc/actions/runs/6919350402). No merge: doesn't seem to help and makes reasoning harder.
- merge the new version of #1499? Should compare to [fb/libm-musl-float](https://github.com/esbmc/esbmc/actions/runs/6919694697), compare to [current](https://github.com/esbmc/esbmc/actions/runs/6919350402). No: we decided that having a correct (or at least one that tries to be correct) implementation is more important.

From discussion elsewhere:
- [x] Rename to `witness.graphml`, see https://gitlab.com/sosy-lab/sv-comp/bench-defs/-/merge_requests/400
- [ ] `--compact-trace` witnesses should not include the `ATOI_MAP` assignments -> make this variable local.
- [x] For valid-cleanup the witness files wrongly contain `CHECK( init(main()), LTL(G valid-free|valid-deref|valid-memtrack) )`
- [ ] The witness files still contain `CHECK( init(main()), LTL(G ! call(__VERIFIER_error())) )` while the unreach-call property [has changed](https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/commit/28369c96c47a133df344c283f0a36d5b761ec3da) to `CHECK( init(main()), LTL(G ! call(reach_error())) )`
- [x] CPAchecker does not like huge numbers like -9223372036854775808 or 18446744073709551615, details from @brcfarias:
  <details>
  
  > So far I have only identified issues in CPAChecker.
  From what I've observed, CPAChecker struggles with handling graphml files containing large numbers.
  Some examples of errors:
  > ```
  > Error: Invalid configuration (Cannot parse witness: The witness automaton provided is invalid! Reason: Cannot parse <data = -9223372036854775808;>) (AutomatonGraphmlParser.getAssumptions, SEVERE)
  >
  > Error: Invalid configuration (Cannot parse witness: The witness automaton provided is invalid! Reason: Cannot parse <combined_mask = 18446744073709551615;>) (AutomatonGraphmlParser.getAssumptions, SEVERE)
  > ```
  > When I change the values to hexadecimal, there is no error, but the verification still ends with 'unknown' for these cases.
  > I've also encountered this type of error in CPAChecker:
  > ```
  > Error: Invalid configuration (Cannot parse witness: The witness automaton provided is invalid! Reason: Cannot parse <iter->element = { .key=(void *)9223934986799808512, .value=(void *)11259136776077632 };>) (AutomatonGraphmlParser.getAssumptions, SEVERE)
  > ```
  
  </details>
  These can be worked around with e.g. `-9223372036854775807 - 1` or by writing them as hexadecimal constants.

We have 2 runs of 30s over the latest `main` versions of the sv-benchmarks, giving an estimate of the expected fluctuation:
```
Statistics:          30471 Files
  correct:           17938
    correct true:    11111
    correct false:    6827
  incorrect:            51
    incorrect true:     19
    incorrect false:    32
  unknown:           12482
  Score:             27929 (max: 51065)

https://github.com/esbmc/esbmc/actions/runs/6747016239
```
and
```
Statistics:          30471 Files
  correct:           17945
    correct true:    11116
    correct false:    6829
  incorrect:            52
    incorrect true:     20
    incorrect false:    32
  unknown:           12474
  Score:             27909 (max: 51065)

https://github.com/esbmc/esbmc/actions/runs/6755039421
```
See https://github.com/esbmc/esbmc/issues/1447#issuecomment-1798475512 for details on incorrect ones.

### CODE REVIEW

**Comment by fbrausse** [comment]
```
The following tests FAILED:
	799 - regression/esbmc/github_60 (Failed)
	1306 - regression/cbmc/01_cbmc_String_Abstraction7 (Failed)
	1456 - regression/cstd/strerror (Failed)
```
The last one, cstd/strerror, is KNOWNBUG, and has only fopen() but no fclose(), so that's fine. First one also has fopen() but no fclose(), however it was supposed to fail with an overflow, now it's successful. That's strange. I also don't understand why the 2nd one, regression/cbmc/01_cbmc_String_Abstraction7, should have been failing in the first place.

**Comment by fbrausse** [comment]
From 2023-11-11 30s on old `main` branch of sv-benchmarks **without `--compact-trace`**:
```
Statistics:          30471 Files
  correct:           17968
    correct true:    11134
    correct false:    6834
  incorrect:            38
    incorrect true:     19
    incorrect false:    19
  unknown:           12465
  Score:             28190 (max: 51065)

https://github.com/esbmc/esbmc/actions/runs/6834053649
```
2023-11-13 30s (updated) main branch of sv-benchmarks **with `--compact-trace`**:
```
Statistics:          31211 Files
  correct:           17871
    correct true:    11231
    correct false:    6640
  incorrect:           106
    incorrect true:     31
    incorrect false:    75
  unknown:           13234
  Score:             26910 (max: 52283)

https://github.com/esbmc/esbmc/actions/runs/6853986122
```
Apples and oranges, I'll start a new run without `--compact-trace` to get either only apples *or* only oranges.

**Comment by fbrausse** [comment]
Difference between 2023-11-11 (default options) and 2023-11-13 (`--compact-trace`).

no-data-race:
- 17x new incorrect
- 31x new unknown (crashes, etc.), these should be investigated

no-overflow:
- 1x new unknown (crash) in c/ldv-linux-4.0-rc1-mav/linux-4.0-rc1---drivers--media--usb--gspca--gspca_main.ko.cil.i

unreach-call:
- new crashes:
  c/array-examples/sorting_bubblesort_2_ground.i
  c/neural-networks/softmax_3_unsafe.c-amalgamation.i
  c/pthread-race-challenges/thread-local-pthread-value.i
  c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--gpu--drm--ttm--ttm.ko-ldv_main5_sequence_infinite_withcheck_stateful.cil.out.i
  c/ldv-consumption/32_7a_cilled_linux-3.8-rc1-32_7a-drivers--net--can--softing--softing.ko-ldv_main0_sequence_infinite_withcheck_stateful.cil.out.i

- 92x new "unknown" due to changed reach_error() usage? All in awc-c-common. This needs ~urgent~ fixing.
- 71x new incorrect:
  - 68x neural networks
  - remaining new incorrect:
    c/pthread-ext/18_read_write_lock-pthread.i
    c/pthread-race-challenges/thread-local-value-dynamic.i
    c/pthread-race-challenges/thread-local-value.i

valid-memsafety:
- 12x new unknown in uthash-2.0.2/uthash_{BER,FNV,JEN,OAT,SAX,SFH}_test6-[12]: `dereference failure: memset of memory segment of size 40 with 44 bytes` (all have verdict true)
- 1x false(valid-deref) -> false(valid-memtrack): c/goblint-regression/09-regions_13-arraycollapse_nr-deref.i

**Comment by fbrausse** [comment]
Let's try the `--compact-trace` comparison again. sv-benchmarks revision f1bf227b2d313c8e28efea5e8364d6347da3f639 (their main branch as of 2023-11-13) with this branch on 00acc9c1ccdb2555d2ecc83272874b7373b6aa36 and **default options**:
```
Statistics:          31211 Files
  correct:           18439
    correct true:    11535
    correct false:    6904
  incorrect:           111
    incorrect true:     33
    incorrect false:    78
  unknown:           12661
  Score:             27670 (max: 52283)

https://github.com/esbmc/esbmc/actions/runs/6862509546
```
Same sv-benchmarks and same ESBMC revision and **--compact-trace**:
```
Statistics:          31211 Files
  correct:           17871
    correct true:    11231
    correct false:    6640
  incorrect:           106
    incorrect true:     31
    incorrect false:    75
  unknown:           13234
  Score:             26910 (max: 52283)

https://github.com/esbmc/esbmc/actions/runs/6853986122
```

**Comment by fbrausse** [comment]
> Let's try the `--compact-trace` comparison again.

It's not too bad, actually. I'll look at it in more detail tomorrow to see whether it might make sense to enable for some properties. In particular it would be good to know whether the validators benefit from that and can compensate for the ~750 points we lose by enabling this option.

@rafaelsamenezes Do you know whether the score printed by benchexec does already include the normalizations Dirk does for the category sizes?

**Comment by fbrausse** [comment]
Between https://github.com/esbmc/esbmc/actions/runs/6883131553 and https://github.com/esbmc/esbmc/actions/runs/6906252342 lots of benchmarks are new. I'm not sure the new ones will be included in SV-COMP'24.

no-overflow:
- 30x new incorrect-false, floating-point multiplication results in infinity; could also be a bug upstream

unreach-call:
- 14x incorrect NNs have been fixed, or now time out
- 5x new tasks are unknown, 1x previously time out now also unknown

valid-memsafety:
- fixed 6x unknowns in ldv-memsafety/memset*
- fixed lots of builtin_unreachable parsing errors
- deterministic uthash now timeout

@rafaelsamenezes I think we should restrict our tests to the `svcomp24-rc.1` tag instead of the `main` branch, if I understood Philipp correctly: `main` is keeping to move forward while from the former Dirk will exclude all tasks touched by MRs, right?

I'll merge the new PRs and start a run on `svcomp24-rc.1` so we don't have new upstream tasks that aren't going to be in SV-COMP'24.

**Comment by rafaelsamenezes** [comment]
> Between https://github.com/esbmc/esbmc/actions/runs/6883131553 and https://github.com/esbmc/esbmc/actions/runs/6906252342 lots of benchmarks are new. I'm not sure the new ones will be included in SV-COMP'24.
> 
> no-overflow:
> - 30x new incorrect-false, floating-point multiplication results in infinity; could also be a bug upstream
> 
> unreach-call:
> - 14x incorrect NNs have been fixed, or now time out
> - 5x new tasks are unknown, 1x previously time out now also unknown
> 
> valid-memsafety:
> - fixed 6x unknowns in ldv-memsafety/memset*
> - fixed lots of builtin_unreachable parsing errors
> - deterministic uthash now timeout
> 
> @rafaelsamenezes I think we should restrict our tests to the `svcomp24-rc.1` tag instead of the `main` branch, if I understood Philipp correctly: `main` is keeping to move forward while from the former Dirk will exclude all tasks touched by MRs, right?
> 
> I'll merge the new PRs and start a run on `svcomp24-rc.1` so we don't have new upstream tasks that aren't going to be in SV-COMP'24.


Indeed, rc.1 will be the one running. Although, some of main fixes may go to the official release. So we should pay attention to what affects us (e.g. the neural networks).

**Comment by fbrausse** [comment]
30s on the sv-benchmarks `svcomp24-rc.1` branch.
Current state of this PR:
```
Statistics:          31649 Files
  correct:           18425
    correct true:    11492
    correct false:    6933
  incorrect:            93
    incorrect true:     33
    incorrect false:    60
  unknown:           13131
  Score:             27901 (max: 53010)

https://github.com/esbmc/esbmc/actions/runs/6919350402
```
With the changes from #1505 ontop:
```
Statistics:          31649 Files
  correct:           18403
    correct true:    11472
    correct false:    6931
  incorrect:            93
    incorrect true:     33
    incorrect false:    60
  unknown:           13153
  Score:             27859 (max: 53010)

https://github.com/esbmc/esbmc/actions/runs/6919691385
```
This PR with musl's float implementations of the functions required for NNs (branch `fb/libm-musl-float`):
```
Statistics:          31649 Files
  correct:           18456
    correct true:    11535
    correct false:    6921
  incorrect:            50
    incorrect true:     32
    incorrect false:    18
  unknown:           13143
  Score:             28679 (max: 53010)

https://github.com/esbmc/esbmc/actions/runs/6919694697
```

**Comment by fbrausse** [comment]
90s run just on the ReachSafety-Floats tasks (which include the NNs), using `--incremental-bmc` instead of `--k-induction`:
Current version of this branch:
```
Statistics:           1076 Files
  correct:             451
    correct true:      392
    correct false:      59
  incorrect:            46
    incorrect true:      1
    incorrect false:    45
  unknown:             579
  Score:                91 (max: 1883)

https://github.com/esbmc/esbmc/actions/runs/6920214487
```
With @emanino's current version #1499:
```
Statistics:           1076 Files
  correct:             434
    correct true:      393
    correct false:      41
  incorrect:             1
    incorrect true:      1
    incorrect false:     0
  unknown:             641
  Score:               795 (max: 1883)

https://github.com/esbmc/esbmc/actions/runs/6926681052
```
With just the NN-required float models from @emanino's #1232:
```
Statistics:           1076 Files
  correct:             427
    correct true:      385
    correct false:      42
  incorrect:             3
    incorrect true:      1
    incorrect false:     2
  unknown:             646
  Score:               748 (max: 1883)

https://github.com/esbmc/esbmc/actions/runs/6927845984
```
The incorrect in both, rough and musl, above are addressed upstream, they're not our bug:
- https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/merge_requests/1491
- https://gitlab.com/sosy-lab/benchmarking/sv-benchmarks/-/merge_requests/1490

**Comment by fbrausse** [comment]
Latest 30s run looks OK except for some of the no-overflow nla-* ones which aren't addressed upstream (yet?):
```
Statistics:          32533 Files
  correct:           18652
    correct true:    11733
    correct false:    6919
  incorrect:            72
    incorrect true:     22
    incorrect false:    50
  unknown:           13809
  Score:             28881 (max: 54621)

https://github.com/esbmc/esbmc/actions/runs/6961415893
```

**Comment by fbrausse** [comment]
For reference, abc9f1bd0dd30b1b1008c1c50cdcdfd7700cde7c (tag `v7.4-svcomp24`) was used to build https://zenodo.org/records/10198805 and is the final version submitted to and merged into fm-tools: https://gitlab.com/sosy-lab/benchmarking/fm-tools/-/merge_requests/170

- esbmc-kind.xml: https://gitlab.com/sosy-lab/sv-comp/bench-defs/-/blob/946214aa74b0a125427cc53400d6e4ae344c172a/benchmark-defs/esbmc-kind.xml
- categories: https://gitlab.com/sosy-lab/sv-comp/bench-defs/-/blob/946214aa74b0a125427cc53400d6e4ae344c172a/benchmark-defs/category-structure.yml
- https://gitlab.com/sosy-lab/sv-comp/sv-comp-2024/-/issues/24
- last pre-run results under https://sv-comp.sosy-lab.org/2024/results/results-verified/ are identified by `esbmc-kind.2023-11-23_23-24-49.results.SV-COMP24_*.xml.bz2.fixed.xml.bz2` for the XMLs (replace `*` by category) and the HTML tables by appending `.table.html` to that

**Comment by lucasccordeiro** [comment]
Many thanks, @fbrausse.
